File: Extensions/Tessa.Extensions.Default.Shared/Workflow/KrProcess/Formatters/ProcessManagementStageTypeFormatter.cs
#nullable enable

using System.Collections.Generic;
using System.Threading.Tasks;
using Tessa.Platform;
using Tessa.Platform.Storage;

namespace Tessa.Extensions.Default.Shared.Workflow.KrProcess.Formatters
{
    /// <summary>
    /// Форматтер типа этапа <see cref="StageTypeDescriptors.ProcessManagementDescriptor"/>.
    /// </summary>
    public sealed class ProcessManagementStageTypeFormatter :
        StageTypeFormatterBase
    {
        #region Base Overrides

        /// <inheritdoc />
        public override ValueTask FormatClientAsync(
            IStageTypeFormatterContext context)
        {
            FormatInternal(
                context,
                context.StageRow.Fields,
                true);

            return ValueTask.CompletedTask;
        }

        /// <inheritdoc />
        public override ValueTask FormatServerAsync(
            IStageTypeFormatterContext context)
        {
            FormatInternal(
                context,
                context.Settings!,
                false);

            return ValueTask.CompletedTask;
        }

        #endregion

        #region Private Methods

        private static void FormatInternal(
            IStageTypeFormatterContext context,
            IDictionary<string, object?> settings,
            bool isClient)
        {
            var sb = StringBuilderHelper.Acquire(256);

            AppendString(
                sb,
                settings.TryGet<string>(KrConstants.KrProcessManagementStageSettingsVirtual.ModeName),
                null,
                true,
                limit: isClient ? DefaultSettingMax : -1);

            var modeID = (ProcessManagementStageTypeMode?) settings.TryGet<int?>(KrConstants.KrProcessManagementStageSettingsVirtual.ModeID);
            switch (modeID)
            {
                case ProcessManagementStageTypeMode.StageMode:
                    var stageName = settings.TryGet<string>(KrConstants.KrProcessManagementStageSettingsVirtual.StageName)?.Trim();

                    if (!string.IsNullOrEmpty(stageName))
                    {
                        AppendString(
                            sb,
                            stageName,
                            null,
                            true,
                            limit: isClient ? DefaultSettingMax : -1);

                        var groupRowName = settings.TryGet<string>(KrConstants.KrProcessManagementStageSettingsVirtual.StageRowGroupName)?.Trim();

                        if (!string.IsNullOrEmpty(groupRowName))
                        {
                            sb.Append(" (");
                            AppendString(
                                sb,
                                groupRowName,
                                null,
                                true,
                                limit: isClient ? DefaultSettingMax : -1,
                                appendNewLine: false);
                            sb.Append(')');
                        }
                    }

                    break;

                case ProcessManagementStageTypeMode.GroupMode:
                    AppendString(
                        sb,
                        settings.TryGet<string>(KrConstants.KrProcessManagementStageSettingsVirtual.StageGroupName),
                        null,
                        true,
                        limit: isClient ? DefaultSettingMax : -1);

                    break;

                case ProcessManagementStageTypeMode.SendSignalMode:
                    AppendString(
                        sb,
                        settings.TryGet<string>(KrConstants.KrProcessManagementStageSettingsVirtual.Signal),
                        null,
                        false,
                        limit: isClient ? DefaultSettingMax : -1);

                    break;
            }

            if (settings.TryGet<bool?>(KrConstants.KrProcessManagementStageSettingsVirtual.ManagePrimaryProcess) == true)
            {
                if (sb.Length > 0)
                {
                    sb.AppendLine();
                }

                sb.Append("{$CardTypes_Controls_ManagePrimaryProcess}");
            }

            context.DisplaySettings = sb.ToStringAndRelease();
        }

        #endregion
    }
}

File: WebClient SDK/src/default/workflow/krProcess/formatters/processManagementStageTypeFormatter.ts
import { extension } from '@tessa/application';
import {
  IKrStageTypeFormatterContext,
  KrStageTypeFormatter,
  KrStageTypeFormattingHelper,
  StageTypeDescriptor,
  StageTypeDescriptors
} from '@tessa/platform';
import { ProcessManagementStageTypeMode } from './../processManagementStageTypeMode';

@extension({ name: 'ProcessManagementStageTypeFormatter' })
export class ProcessManagementStageTypeFormatter extends KrStageTypeFormatter {
  //#region constants and static fields

  private static readonly _managePrimaryProcess = KrStageTypeFormattingHelper.formatPlainColumnName(
    'KrProcessManagementStageSettingsVirtual',
    'ManagePrimaryProcess'
  );
  private static readonly _modeId = KrStageTypeFormattingHelper.formatPlainColumnName(
    'KrProcessManagementStageSettingsVirtual',
    'ModeID'
  );
  private static readonly _modeName = KrStageTypeFormattingHelper.formatPlainColumnName(
    'KrProcessManagementStageSettingsVirtual',
    'ModeName'
  );
  private static readonly _stageName = KrStageTypeFormattingHelper.formatPlainColumnName(
    'KrProcessManagementStageSettingsVirtual',
    'StageName'
  );
  private static readonly _groupName = KrStageTypeFormattingHelper.formatPlainColumnName(
    'KrProcessManagementStageSettingsVirtual',
    'StageGroupName'
  );
  private static readonly _groupRowName = KrStageTypeFormattingHelper.formatPlainColumnName(
    'KrProcessManagementStageSettingsVirtual',
    'StageRowGroupName'
  );
  private static readonly _signal = KrStageTypeFormattingHelper.formatPlainColumnName(
    'KrProcessManagementStageSettingsVirtual',
    'Signal'
  );

  //#endregion

  //#region base overrides

  descriptors(): ReadonlyArray<StageTypeDescriptor> {
    return [StageTypeDescriptors.processManagementDescriptor];
  }

  format(context: IKrStageTypeFormatterContext): void {
    const builder = { text: '' };

    this.appendString(
      builder,
      context.stageRow.tryGetString(ProcessManagementStageTypeFormatter._modeName),
      '',
      true,
      false,
      KrStageTypeFormatter.defaultSettingMax
    );

    switch (context.stageRow.tryGet(ProcessManagementStageTypeFormatter._modeId)) {
      case ProcessManagementStageTypeMode.StageMode:
        const stageName = context.stageRow.tryGetString(
          ProcessManagementStageTypeFormatter._stageName
        );

        if (stageName) {
          this.appendString(
            builder,
            stageName,
            '',
            true,
            false,
            KrStageTypeFormatter.defaultSettingMax
          );

          const groupRowName = context.stageRow.tryGet<string>(
            ProcessManagementStageTypeFormatter._groupRowName
          );

          if (groupRowName) {
            builder.text += ' (';
            this.appendString(
              builder,
              groupRowName,
              '',
              true,
              false,
              KrStageTypeFormatter.defaultSettingMax,
              false
            );
            builder.text += ')';
          }
        }
        break;
      case ProcessManagementStageTypeMode.GroupMode:
        this.appendString(
          builder,
          context.stageRow.tryGetString(ProcessManagementStageTypeFormatter._groupName),
          '',
          true,
          false,
          KrStageTypeFormatter.defaultSettingMax
        );
        break;
      case ProcessManagementStageTypeMode.SendSignalMode:
        this.appendString(
          builder,
          context.stageRow.tryGetString(ProcessManagementStageTypeFormatter._signal),
          '',
          true,
          false,
          KrStageTypeFormatter.defaultSettingMax
        );
        break;
    }

    if (context.stageRow.tryGetBoolean(ProcessManagementStageTypeFormatter._managePrimaryProcess)) {
      if (builder.text.length > 0) {
        builder.text += '\n';
      }

      builder.text += '{$CardTypes_Controls_ManagePrimaryProcess}\n';
    }

    context.displaySettings = builder.text;
  }

  //#endregion
}
